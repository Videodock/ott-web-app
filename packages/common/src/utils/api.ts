import type { CommonResponse } from '@inplayer-org/inplayer.js';

import type { InPlayerError } from '../../types/inplayer';

export class ApiError extends Error {
  code: number;
  message: string;

  constructor(message = '', code: number) {
    super(message);
    this.name = 'ApiError';
    this.message = message;
    this.code = code;
  }
}

/**
 * Get data
 * @param response
 */
export const getDataOrThrow = async (response: Response) => {
  const data = await response.json();

  if (!response.ok) {
    const message = `Request '${response.url}' failed with ${response.status}`;
    const apiMessage = data && typeof data === 'object' && 'message' in data && typeof data.message === 'string' ? data.message : undefined;

    throw new ApiError(apiMessage || message, response.status || 500);
  }

  return data;
};

export const getCommonResponseData = (response: { data: CommonResponse }) => {
  const { code, message } = response.data;
  if (code !== 200) {
    throw new Error(message);
  }
  return {
    errors: [],
    responseData: {
      message,
      code,
    },
  };
};

export const isCommonError = (error: unknown): error is InPlayerError => {
  return (
    typeof error === 'object' &&
    error !== null &&
    'response' in error &&
    typeof (error as InPlayerError).response?.data?.code === 'number' &&
    typeof (error as InPlayerError).response?.data?.message === 'string'
  );
};
