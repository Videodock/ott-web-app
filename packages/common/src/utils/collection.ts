import type { CustomFormField, ConsentsValue } from '../../types/account';
import type { Config } from '../../types/config';
import type { GenericFormValues } from '../../types/form';
import type { Playlist, PlaylistItem } from '../../types/playlist';
import { CARD_ASPECT_RATIOS } from '../constants';

export type PosterAspectRatio = (typeof CARD_ASPECT_RATIOS)[number];

const getFiltersFromConfig = (config: Config, playlistId: string | undefined): string[] => {
  const menuItem = config.menu.find((item) => item.contentId === playlistId);
  const filters = menuItem?.filterTags?.split(',').filter(Boolean);

  return filters || [];
};

const filterPlaylist = (playlist: Playlist, filter: string) => {
  if (filter === '') return playlist;

  return {
    ...playlist,
    playlist: playlist.playlist.filter(({ tags }) => (tags ? tags.split(',').includes(filter) : false)),
  };
};

const chunk = <T>(input: T[], size: number) => {
  return input?.reduce((arr: T[][], item, idx: number) => {
    return idx % size === 0 ? [...arr, [item]] : [...arr.slice(0, -1), [...arr.slice(-1)[0], item]];
  }, []);
};

const findPlaylistImageForWidth = (playlistItem: PlaylistItem, width: number): string =>
  playlistItem.images.find((img) => img.width === width)?.src || playlistItem.image;

const generatePlaylistPlaceholder = (playlistLength: number = 15): Playlist => ({
  title: '',
  playlist: new Array(playlistLength).fill({}).map(
    (_value, index) =>
      ({
        description: '',
        duration: 0,
        feedid: '',
        image: '',
        images: [],
        cardImage: '',
        backgroundImage: '',
        channelLogoImage: '',
        link: '',
        genre: '',
        mediaid: `placeholder_${index}`,
        pubdate: 0,
        rating: '',
        sources: [],
        tags: '',
        title: '',
        tracks: [],
      } as PlaylistItem),
  ),
});

const formatConsentValues = (publisherConsents: CustomFormField[] | null = [], customerConsents: ConsentsValue[] | null = []) => {
  if (!publisherConsents || !customerConsents) {
    return {};
  }

  return publisherConsents.reduce((acc, publisherConsent) => {
    const consent = customerConsents?.find((customerConsent) => customerConsent.name === publisherConsent.name);

    if (consent) {
      acc[publisherConsent.name] = consent.state === 'accepted';
    }

    return acc;
  }, {} as Record<string, boolean>);
};

const formatConsents = (publisherConsents: CustomFormField[] | null = [], customerConsents: ConsentsValue[] | null = []) => {
  if (!publisherConsents || !customerConsents) {
    return {};
  }
  const values: Record<string, boolean> = {};
  publisherConsents?.forEach((publisherConsent) => {
    if (customerConsents?.find((customerConsent) => customerConsent.name === publisherConsent.name && customerConsent.state === 'accepted')) {
      values[publisherConsent.name] = true;
    }
  });

  return values;
};

const extractConsentValues = (consents?: CustomFormField[]) => {
  const values: Record<string, string | boolean> = {};

  if (!consents) {
    return values;
  }

  consents?.forEach((consent) => {
    values[consent.name] = consent.type === 'checkbox' ? consent.enabledByDefault === true : consent.defaultValue ?? '';
  });

  return values;
};

const formatConsentsFromValues = (publisherConsents: CustomFormField[] | null, values?: GenericFormValues) => {
  const consents: ConsentsValue[] = [];

  if (!publisherConsents || !values) return consents;

  publisherConsents.forEach((consent) => {
    consents.push({
      name: consent.name,
      version: consent.version || '1',
      state: values[consent.name] ? 'accepted' : 'declined',
    });
  });

  return consents;
};

const checkConsentsFromValues = (publisherConsents: CustomFormField[], consents: Record<string, string | boolean>) => {
  const customerConsents: ConsentsValue[] = [];
  const consentsErrors: string[] = [];

  if (!publisherConsents || !consents) return { customerConsents, consentsErrors };

  publisherConsents.forEach((consent) => {
    if (consent.required && !consents[consent.name]) {
      consentsErrors.push(consent.name);
    }

    customerConsents.push({
      name: consent.name,
      version: consent.version || '1',
      state: consents[consent.name] ? 'accepted' : 'declined',
    });
  });

  return { customerConsents, consentsErrors };
};

const deepCopy = (obj: unknown) => {
  if (Array.isArray(obj) || (typeof obj === 'object' && obj !== null)) {
    return JSON.parse(JSON.stringify(obj));
  }
  return obj;
};

const parseAspectRatio = (input: unknown) => {
  if (typeof input === 'string' && (CARD_ASPECT_RATIOS as readonly string[]).includes(input)) return input as PosterAspectRatio;
};

const parseTilesDelta = (posterAspect?: PosterAspectRatio) => {
  if (!posterAspect) {
    return 0;
  }

  const parts = posterAspect.split(':');

  return parts.length === 2 ? Math.floor(parseInt(parts[1]) / parseInt(parts[0])) : 0;
};

export {
  getFiltersFromConfig,
  filterPlaylist,
  chunk,
  findPlaylistImageForWidth,
  generatePlaylistPlaceholder,
  formatConsentValues,
  formatConsents,
  formatConsentsFromValues,
  extractConsentValues,
  checkConsentsFromValues,
  deepCopy,
  parseAspectRatio,
  parseTilesDelta,
};
