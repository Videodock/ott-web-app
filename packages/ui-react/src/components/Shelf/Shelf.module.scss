.shelf {
  margin: 0 0;
  color: var(--primary-color);
  font-family: var(--body-alt-font-family);

  &:hover,
  &:focus-within {
    .chevron:not(:disabled) {
      opacity: 1;
    }
  }
}

.title {
  width: 100%;
  min-height: 28px;
  margin-bottom: 12px;
  overflow: hidden;
  color: var(--card-color);
  font-family: var(--body-alt-font-family);
  font-weight: var(--body-font-weight-bold);
  font-size: 24px;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.slider {
  position: static;
  overflow: visible;

  :global(.TileSlider-list) {
    li > a {
      white-space: initial;
    }

    li {
      opacity: 1;
      transition: opacity 0.1s ease;

      &:global(.TileSlider--hidden) {
        opacity: 0.5;
      }
    }
  }
}

.chevron {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  align-content: center;
  width: 44px;
  height: 44px;
  outline-color: var(--highlight-color, white);
  cursor: pointer;
  transition: transform 0.3s ease-out, opacity 0.3s ease-out;
  appearance: none;

  > svg {
    width: 30px;
    height: 30px;
  }

  &:disabled {
    opacity: 0.3;
    pointer-events: none;
  }

  &:hover {
    transform: scale(1.2);
  }
}

.dots {
  position: relative;
  display: flex;
  justify-content: center;
  width: 100%;
  margin-top: 12px;
}

.dot {
  display: inline-block;
  width: 10px;
  height: 10px;
  margin: 0 5px;
  background-color: rgba(254, 254, 254, 0.2);
  border-radius: 50%;
  transition: all 200ms ease;

  &.active {
    background-color: var(--primary-color);
    transform: scale(1.1);
  }
}

.error {
  color: var(--card-color);
  font-family: var(--body-alt-font-family);
}
