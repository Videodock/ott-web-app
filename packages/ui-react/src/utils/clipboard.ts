export const copyToClipboard = (value: string): void => {
  const inputElement = document.createElement('input');
  inputElement.style.zIndex = '-10';
  inputElement.style.position = 'absolute';
  inputElement.style.left = '-1000';
  inputElement.value = value;
  document.body.appendChild(inputElement);
  inputElement.select();
  document.execCommand('copy');
  inputElement.blur();
  document.body.removeChild(inputElement);
};
