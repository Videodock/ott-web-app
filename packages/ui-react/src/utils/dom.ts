export const canUseDOM = !!(typeof window !== 'undefined' && window.document && window.document.createElement);

let size: number;

export default function scrollbarSize(recalc?: boolean) {
  if ((!size && size !== 0) || recalc) {
    if (canUseDOM) {
      const scrollDiv = document.createElement('div');
      scrollDiv.style.position = 'absolute';
      scrollDiv.style.top = '-9999px';
      scrollDiv.style.width = '50px';
      scrollDiv.style.height = '50px';
      scrollDiv.style.overflow = 'scroll';
      document.body.appendChild(scrollDiv);
      size = scrollDiv.offsetWidth - scrollDiv.clientWidth;
      document.body.removeChild(scrollDiv);
    }
  }
  return size;
}

export const addScript = (src: string): Promise<void> => {
  return new Promise((resolve) => {
    const script: HTMLScriptElement = document.createElement('script');
    script.type = 'text/javascript';
    script.async = true;
    script.src = src;
    script.onload = () => resolve();
    script.onerror = (error) => {
      console.info('Error loading external script', error);
      throw new Error('Error loading external script');
    };
    document.body.appendChild(script);
  });
};
