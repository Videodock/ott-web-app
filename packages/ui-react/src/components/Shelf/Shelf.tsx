import React, { useCallback, useState } from 'react';
import classNames from 'classnames';
import { useTranslation } from 'react-i18next';
import type { Playlist, PlaylistItem } from '@jwp/ott-common/types/playlist';
import type { AccessModel, ContentType } from '@jwp/ott-common/types/config';
import { isLocked } from '@jwp/ott-common/src/utils/entitlements';
import { mediaURL } from '@jwp/ott-common/src/utils/urlFormatting';
import { PersonalShelf } from '@jwp/ott-common/src/constants';
import ChevronLeft from '@jwp/ott-theme/assets/icons/chevron_left.svg?react';
import ChevronRight from '@jwp/ott-theme/assets/icons/chevron_right.svg?react';
import useBreakpoint, { Breakpoint, type Breakpoints } from '@jwp/ott-ui-react/src/hooks/useBreakpoint';

import TileDock from '../TileDock/TileDock';
import Card from '../Card/Card';
import Icon from '../Icon/Icon';

import styles from './Shelf.module.scss';

export const tileBreakpoints: Breakpoints = {
  [Breakpoint.xs]: 1,
  [Breakpoint.sm]: 2,
  [Breakpoint.md]: 3,
  [Breakpoint.lg]: 4,
  [Breakpoint.xl]: 5,
};

export const featuredTileBreakpoints: Breakpoints = {
  [Breakpoint.xs]: 1,
  [Breakpoint.sm]: 1,
  [Breakpoint.md]: 1,
  [Breakpoint.lg]: 1,
  [Breakpoint.xl]: 1,
};

export type ShelfProps = {
  playlist: Playlist;
  type: ContentType;
  onCardHover?: (playlistItem: PlaylistItem) => void;
  watchHistory?: { [key: string]: number };
  enableTitle?: boolean;
  enableCardTitles?: boolean;
  featured?: boolean;
  loading?: boolean;
  error?: unknown;
  title?: string;
  accessModel: AccessModel;
  isLoggedIn: boolean;
  hasSubscription: boolean;
  posterAspect?: string;
  visibleTilesDelta?: number;
};

const Shelf = ({
  playlist,
  type,
  onCardHover,
  title,
  watchHistory,
  featured = false,
  loading = false,
  error = null,
  accessModel,
  isLoggedIn,
  hasSubscription,
  posterAspect,
  visibleTilesDelta = 0,
}: ShelfProps) => {
  const breakpoint: Breakpoint = useBreakpoint();
  const { t } = useTranslation('common');
  const [didSlideBefore, setDidSlideBefore] = useState(false);
  const tilesToShow: number = (featured ? featuredTileBreakpoints[breakpoint] : tileBreakpoints[breakpoint]) + visibleTilesDelta;

  const renderTile = useCallback(
    (item: PlaylistItem, isInView: boolean) => {
      const url = mediaURL({ media: item, playlistId: playlist.feedid, play: type === PersonalShelf.ContinueWatching });
      return (
        <Card
          key={item.mediaid}
          progress={watchHistory ? watchHistory[item.mediaid] : undefined}
          onHover={typeof onCardHover === 'function' ? () => onCardHover(item) : undefined}
          featured={featured}
          disabled={!isInView}
          loading={loading}
          isLocked={isLocked(accessModel, isLoggedIn, hasSubscription, item)}
          posterAspect={posterAspect}
          item={item}
          url={url}
        />
      );
    },
    [watchHistory, onCardHover, featured, loading, accessModel, isLoggedIn, hasSubscription, posterAspect, playlist.feedid, type],
  );

  const renderRightControl = useCallback(
    (doSlide: () => void) => (
      <div
        className={styles.chevron}
        role="button"
        tabIndex={0}
        aria-label={t('slide_right')}
        onKeyDown={(event: React.KeyboardEvent) => (event.key === 'Enter' || event.key === ' ') && handleSlide(doSlide)}
        onClick={() => handleSlide(doSlide)}
      >
        <Icon icon={ChevronRight} />
      </div>
    ),
    [t],
  );

  const renderLeftControl = useCallback(
    (doSlide: () => void) => (
      <div
        className={classNames(styles.chevron, {
          [styles.disabled]: !didSlideBefore,
        })}
        role="button"
        tabIndex={didSlideBefore ? 0 : -1}
        aria-label={t('slide_left')}
        onKeyDown={(event: React.KeyboardEvent) => (event.key === 'Enter' || event.key === ' ') && handleSlide(doSlide)}
        onClick={() => handleSlide(doSlide)}
      >
        <Icon icon={ChevronLeft} />
      </div>
    ),
    [didSlideBefore, t],
  );

  const renderPaginationDots = (index: number, pageIndex: number) => (
    <span key={pageIndex} className={classNames(styles.dot, { [styles.active]: index === pageIndex })} />
  );

  const handleSlide = (doSlide: () => void): void => {
    setDidSlideBefore(true);
    doSlide();
  };

  if (error || !playlist?.playlist) return <h2 className={styles.error}>Could not load items</h2>;

  return (
    <div className={classNames(styles.shelf)}>
      {!featured ? <h2 className={classNames(styles.title)}>{title || playlist.title}</h2> : null}
      <TileDock<PlaylistItem>
        items={playlist.playlist}
        tilesToShow={tilesToShow}
        wrapWithEmptyTiles={featured && playlist.playlist.length === 1}
        cycleMode={'restart'}
        showControls={!matchMedia('(hover: none)').matches && !loading}
        showDots={featured}
        transitionTime={'0.3s'}
        spacing={8}
        renderLeftControl={renderLeftControl}
        renderRightControl={renderRightControl}
        renderPaginationDots={renderPaginationDots}
        renderTile={renderTile}
      />
    </div>
  );
};

export default Shelf;
