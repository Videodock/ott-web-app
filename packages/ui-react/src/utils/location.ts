import { createURL, type QueryParamsArg } from '@jwp/ott-common/src/utils/urlFormatting';
import type { Location } from 'react-router-dom';

import type { AccountModals } from '../containers/AccountModal/AccountModal';

export const createURLFromLocation = (location: Location, queryParams: QueryParamsArg = {}) => {
  return createURL(`${location.pathname}${location.search || ''}`, queryParams);
};

export const modalURLFromLocation = (location: Location, u: keyof AccountModals | null, queryParams?: QueryParamsArg) => {
  return createURL(`${location.pathname}${location.search || ''}`, { u, ...queryParams });
};
