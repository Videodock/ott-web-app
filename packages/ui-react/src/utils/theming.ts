import type { Config } from '@jwp/ott-common/types/config';
import { calculateContrastColor } from '@jwp/ott-common/src/utils/common';
import env from '@jwp/ott-common/src/env';

export const setThemingVariables = (config: Config) => {
  const root = document.querySelector(':root') as HTMLElement;
  const { backgroundColor, highlightColor, headerBackground } = config.styling || {};
  const bodyFont = env.APP_BODY_FONT;
  const bodyAltFont = env.APP_BODY_ALT_FONT;

  if (!root) return;

  if (backgroundColor) {
    root.style.setProperty('--body-background-color', backgroundColor);
    root.style.setProperty('--background-contrast-color', calculateContrastColor(backgroundColor));
  }
  if (highlightColor) {
    root.style.setProperty('--highlight-color', highlightColor);
    root.style.setProperty('--highlight-contrast-color', calculateContrastColor(highlightColor));
  }
  if (headerBackground) {
    root.style.setProperty('--header-background', headerBackground);
    root.style.setProperty('--header-contrast-color', calculateContrastColor(headerBackground));
  }

  if (bodyFont) {
    root.style.setProperty('--body-font-family', bodyFont);
  }
  if (bodyAltFont) {
    root.style.setProperty('--body-alt-font-family', bodyAltFont);
  }
};
