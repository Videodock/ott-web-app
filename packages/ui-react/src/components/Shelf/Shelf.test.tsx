import React from 'react';
import { axe } from 'vitest-axe';

import { renderWithRouter } from '../../../test/utils';

import Shelf from './Shelf';

const playlist = {
  playlist: [
    {
      mediaid: '1234abcd',
      title: 'Movie 1',
      description: 'This is a move One',
      duration: 30,
      feedid: 'ABCDEFGH',
      image: '',
      images: [],
      link: '',
      pubdate: 30000,
      sources: [],
    },
    {
      mediaid: 'aaaaaaaa',
      title: 'Movie 2',
      description: 'This is a move two',
      duration: 31,
      feedid: 'ABCDEFGH',
      image: '',
      images: [],
      link: '',
      pubdate: 30000,
      sources: [],
    },
    {
      mediaid: '12332123',
      title: 'Third movie',
      description: 'Shrek the Third',
      duration: 34,
      feedid: 'ABCDEFGH',
      image: '',
      images: [],
      link: '',
      pubdate: 30004,
      sources: [],
    },
    {
      mediaid: 'ddeeddee',
      title: 'Last playlist item',
      description: 'Indiana Jones and the Last Crusade',
      duration: 1232,
      feedid: 'ABCDEFGH',
      image: '',
      images: [],
      link: '',
      pubdate: 30001,
      sources: [],
    },
  ],
  title: 'My Playlist',
};

describe('Shelf Component tests', () => {
  test('Regular shelf', () => {
    const { container } = renderWithRouter(
      <Shelf
        title="Test Shelf"
        type={'playlist'}
        accessModel={'AVOD'}
        hasSubscription={true}
        isLoggedIn={true}
        playlist={playlist}
        enableCardTitles
        enableTitle
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test('Featured shelf', () => {
    const { container } = renderWithRouter(
      <Shelf
        title="Featured Shelf"
        type={'playlist'}
        accessModel={'AUTHVOD'}
        hasSubscription={true}
        isLoggedIn={true}
        playlist={playlist}
        featured
        enableCardTitles
        enableTitle
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test('WCAG 2.2 (AA) compliant', async () => {
    const { container } = renderWithRouter(
      <>
        <h2>Regular shelf</h2>
        <Shelf
          title="Test Shelf"
          type={'playlist'}
          accessModel={'AVOD'}
          hasSubscription={true}
          isLoggedIn={true}
          playlist={playlist}
          enableCardTitles
          enableTitle
        />
        <h2>Featured shelf</h2>
        <Shelf
          title="Featured Shelf"
          type={'playlist'}
          accessModel={'AUTHVOD'}
          hasSubscription={true}
          isLoggedIn={true}
          playlist={playlist}
          featured
          enableCardTitles
          enableTitle
        />
      </>,
    );

    expect(await axe(container, { runOnly: ['wcag21a', 'wcag21aa', 'wcag22aa'] })).toHaveNoViolations();
  });
});
