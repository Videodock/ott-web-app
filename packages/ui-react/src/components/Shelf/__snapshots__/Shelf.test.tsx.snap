// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shelf Component tests > Featured shelf 1`] = `
<div>
  <div
    class="_shelf_81910b"
  >
    <div
      class="_tileDock_22abb0"
    >
      <div
        class="_leftControl_22abb0"
      >
        <div
          aria-label="slide_left"
          class="_chevron_81910b _disabled_81910b"
          role="button"
          tabindex="-1"
        >
          <svg
            aria-hidden="true"
            class="_icon_585b29"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
            />
          </svg>
        </div>
      </div>
      <ul
        style="transform: translate3d(-200%, 0, 0); margin-left: -4px; margin-right: -4px;"
      >
        <li
          aria-hidden="true"
          aria-posinset="3"
          aria-setsize="4"
          class="_notInView_22abb0"
          style="width: 100%; padding-left: 4px; padding-right: 4px; box-sizing: border-box; transition: opacity .2s ease-in 0s;"
          tabindex="-1"
        >
          <a
            class="_card_d75732 _featured_d75732 _disabled_d75732"
            data-testid="Third movie"
            href="/m/12332123/third-movie"
            role="button"
            tabindex="-1"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
          </a>
        </li>
        <li
          aria-hidden="true"
          aria-posinset="4"
          aria-setsize="4"
          class="_notInView_22abb0"
          style="width: 100%; padding-left: 4px; padding-right: 4px; box-sizing: border-box; transition: opacity .2s ease-in 0s;"
          tabindex="-1"
        >
          <a
            class="_card_d75732 _featured_d75732 _disabled_d75732"
            data-testid="Last playlist item"
            href="/m/ddeeddee/last-playlist-item"
            role="button"
            tabindex="-1"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
          </a>
        </li>
        <li
          aria-hidden="false"
          aria-posinset="1"
          aria-setsize="4"
          class=""
          style="width: 100%; padding-left: 4px; padding-right: 4px; box-sizing: border-box;"
          tabindex="-1"
        >
          <a
            class="_card_d75732 _featured_d75732"
            data-testid="Movie 1"
            href="/m/1234abcd/movie-1"
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <h3
                  class="_title_d75732"
                >
                  Movie 1
                </h3>
                <div
                  class="_tags_d75732"
                >
                  <div
                    class="_tag_d75732"
                  >
                    1 min
                  </div>
                </div>
              </div>
            </div>
          </a>
        </li>
        <li
          aria-hidden="true"
          aria-posinset="2"
          aria-setsize="4"
          class="_notInView_22abb0"
          style="width: 100%; padding-left: 4px; padding-right: 4px; box-sizing: border-box; transition: opacity .2s ease-in 0s;"
          tabindex="-1"
        >
          <a
            class="_card_d75732 _featured_d75732 _disabled_d75732"
            data-testid="Movie 2"
            href="/m/aaaaaaaa/movie-2"
            role="button"
            tabindex="-1"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
          </a>
        </li>
        <li
          aria-hidden="true"
          aria-posinset="3"
          aria-setsize="4"
          class="_notInView_22abb0"
          style="width: 100%; padding-left: 4px; padding-right: 4px; box-sizing: border-box; transition: opacity .2s ease-in 0s;"
          tabindex="-1"
        >
          <a
            class="_card_d75732 _featured_d75732 _disabled_d75732"
            data-testid="Third movie"
            href="/m/12332123/third-movie"
            role="button"
            tabindex="-1"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
          </a>
        </li>
      </ul>
      <div
        class="_rightControl_22abb0"
      >
        <div
          aria-label="slide_right"
          class="_chevron_81910b"
          role="button"
          tabindex="0"
        >
          <svg
            aria-hidden="true"
            class="_icon_585b29"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="_dots_22abb0"
    >
      <span
        class="_dot_81910b _active_81910b"
      />
      <span
        class="_dot_81910b"
      />
      <span
        class="_dot_81910b"
      />
      <span
        class="_dot_81910b"
      />
    </div>
  </div>
</div>
`;

exports[`Shelf Component tests > Regular shelf 1`] = `
<div>
  <div
    class="_shelf_81910b"
  >
    <h2
      class="_title_81910b"
    >
      Test Shelf
    </h2>
    <div
      class="_tileDock_22abb0"
    >
      <ul
        style="transform: translate3d(0%, 0, 0); margin-left: -4px; margin-right: -4px;"
      >
        <li
          aria-hidden="false"
          aria-posinset="1"
          aria-setsize="4"
          class=""
          style="width: 20%; padding-left: 4px; padding-right: 4px; box-sizing: border-box;"
          tabindex="-1"
        >
          <a
            class="_card_d75732"
            data-testid="Movie 1"
            href="/m/1234abcd/movie-1"
            role="button"
            tabindex="0"
          >
            <div
              class="_titleContainer_d75732"
            >
              <h3
                class="_title_d75732"
              >
                Movie 1
              </h3>
            </div>
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                >
                  <div
                    class="_tag_d75732"
                  >
                    1 min
                  </div>
                </div>
              </div>
            </div>
          </a>
        </li>
        <li
          aria-hidden="false"
          aria-posinset="2"
          aria-setsize="4"
          class=""
          style="width: 20%; padding-left: 4px; padding-right: 4px; box-sizing: border-box;"
          tabindex="-1"
        >
          <a
            class="_card_d75732"
            data-testid="Movie 2"
            href="/m/aaaaaaaa/movie-2"
            role="button"
            tabindex="0"
          >
            <div
              class="_titleContainer_d75732"
            >
              <h3
                class="_title_d75732"
              >
                Movie 2
              </h3>
            </div>
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                >
                  <div
                    class="_tag_d75732"
                  >
                    1 min
                  </div>
                </div>
              </div>
            </div>
          </a>
        </li>
        <li
          aria-hidden="false"
          aria-posinset="3"
          aria-setsize="4"
          class=""
          style="width: 20%; padding-left: 4px; padding-right: 4px; box-sizing: border-box;"
          tabindex="-1"
        >
          <a
            class="_card_d75732"
            data-testid="Third movie"
            href="/m/12332123/third-movie"
            role="button"
            tabindex="0"
          >
            <div
              class="_titleContainer_d75732"
            >
              <h3
                class="_title_d75732"
              >
                Third movie
              </h3>
            </div>
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                >
                  <div
                    class="_tag_d75732"
                  >
                    1 min
                  </div>
                </div>
              </div>
            </div>
          </a>
        </li>
        <li
          aria-hidden="false"
          aria-posinset="4"
          aria-setsize="4"
          class=""
          style="width: 20%; padding-left: 4px; padding-right: 4px; box-sizing: border-box;"
          tabindex="-1"
        >
          <a
            class="_card_d75732"
            data-testid="Last playlist item"
            href="/m/ddeeddee/last-playlist-item"
            role="button"
            tabindex="0"
          >
            <div
              class="_titleContainer_d75732"
            >
              <h3
                class="_title_d75732"
              >
                Last playlist item
              </h3>
            </div>
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                >
                  <div
                    class="_tag_d75732"
                  >
                    21 min
                  </div>
                </div>
              </div>
            </div>
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
